corehq/apps/hqcase/api/get_list.py

```diff
@@ -0,0 +1,111 @@
+import datetime
+from itertools import chain
+
+from dateutil.parser import parse
+
+from dimagi.utils.parsing import FALSE_STRINGS
+
+from corehq.apps.case_search.filter_dsl import (
+    CaseFilterError,
+    build_filter_from_xpath,
+)
+from corehq.apps.es import case_search
+from corehq.apps.es import cases as case_es
+
+from .core import UserError, serialize_es_case
+
+DEFAULT_PAGE_SIZE = 20
+MAX_PAGE_SIZE = 5000
+CUSTOM_PROPERTY_PREFIX = 'property.'
+
+
+def _to_boolean(val):
+    return not (val == '' or val.lower() in FALSE_STRINGS)
+
+
+def _to_int(val, param_name):
+    try:
+        return int(val)
+    except ValueError:
+        raise UserError(f"'{val}' is not a valid value for '{param_name}'")
+
+
+def _make_date_filter(date_filter, param):
+
+    def filter_fn(val):
+        try:
+            # If it's only a date, don't turn it into a datetime
+            val = datetime.datetime.strptime(val, '%Y-%m-%d').date()
+        except ValueError:
+            try:
+                val = parse(val)
+            except ValueError:
+                raise UserError(f"Cannot parse datetime '{val}'")
+        return date_filter(**{param: val})
+
+    return filter_fn
+
+
+def _to_date_filters(field, date_filter):
+    return [
+        (f'{field}.gt', _make_date_filter(date_filter, 'gt')),
+        (f'{field}.gte', _make_date_filter(date_filter, 'gte')),
+        (f'{field}.lte', _make_date_filter(date_filter, 'lte')),
+        (f'{field}.lt', _make_date_filter(date_filter, 'lt')),
+        (f'{field}.eq', _make_date_filter(date_filter, 'eq')),
+    ]
+
+
+FILTERS = {
+    'external_id': case_search.external_id,
+    'case_type': case_es.case_type,
+    'owner_id': case_es.owner,
+    'case_name': case_es.case_name,
+    'closed': lambda val: case_es.is_closed(_to_boolean(val)),
+}
+FILTERS.update(chain(*[
+    _to_date_filters('last_modified', case_es.modified_range),
+    _to_date_filters('server_last_modified', case_es.server_modified_range),
+    _to_date_filters('date_opened', case_es.opened_range),
+    _to_date_filters('date_closed', case_es.closed_range),
+    _to_date_filters('indexed_on', case_search.indexed_on),
+]))
+
+
+def get_list(domain, params):
+    start = _to_int(params.pop('offset', 0), 'offset')
+    page_size = _to_int(params.pop('limit', DEFAULT_PAGE_SIZE), 'limit')
+    if page_size > MAX_PAGE_SIZE:
+        raise UserError(f"You cannot request more than {MAX_PAGE_SIZE} cases per request.")
```

Comment:
Not blocking this PR but I was thinking more about the pagination last night and wondering if we should even support this kind of pagination on these APIs since we know they suffer from the deep pagination problem. The alternative would be to use cursor pagination: https://www.django-rest-framework.org/api-guide/pagination/#cursorpagination, https://cra.mr/2011/03/08/building-cursors-for-the-disqus-api/

Reply:
interesting - I'll read up on that, thanks for the links